src/Runner.Worker/StepsRunner.cs

```diff
@@ -98,124 +98,139 @@ public async Task RunAsync(IExecutionContext jobContext)
                         envContext[pair.Key] = new StringContextData(pair.Value ?? string.Empty);
                     }
 
+                    bool evaluateStepEnvFailed = false;
                     if (step is IActionRunner actionStep)
                     {
                         // Set GITHUB_ACTION
                         step.ExecutionContext.SetGitHubContext("action", actionStep.Action.Name);
 
-                        // Evaluate and merge action's env block to env context
-                        var templateEvaluator = step.ExecutionContext.ToPipelineTemplateEvaluator();
-                        var actionEnvironment = templateEvaluator.EvaluateStepEnvironment(actionStep.Action.Environment, step.ExecutionContext.ExpressionValues, step.ExecutionContext.ExpressionFunctions, VarUtil.EnvironmentVariableKeyComparer);
-                        foreach (var env in actionEnvironment)
+                        try
                         {
-                            envContext[env.Key] = new StringContextData(env.Value ?? string.Empty);
+                            // Evaluate and merge action's env block to env context
+                            var templateEvaluator = step.ExecutionContext.ToPipelineTemplateEvaluator();
+                            var actionEnvironment = templateEvaluator.EvaluateStepEnvironment(actionStep.Action.Environment, step.ExecutionContext.ExpressionValues, step.ExecutionContext.ExpressionFunctions, VarUtil.EnvironmentVariableKeyComparer);
+                            foreach (var env in actionEnvironment)
+                            {
+                                envContext[env.Key] = new StringContextData(env.Value ?? string.Empty);
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            // fail the step since there is an evaluate error.
+                            Trace.Info("Caught exception from expression for step.env");
+                            evaluateStepEnvFailed = true;
+                            step.ExecutionContext.Error(ex);
+                            CompleteStep(step, nextStep, TaskResult.Failed);
                         }
                     }
 
-                    try
+                    if (!evaluateStepEnvFailed)
```

Comment:
the diff is bad, the real change is to skip execute the step when `EvaluateStepEnvironment` failed, and merge step result with job result.

Reply:
did you consider merging this with current local vars for tracking whether error when evaluating `if` condition?

Reply:
i dont know whether possible (something to consider whether it would simplify)

Reply:
looking further... ignore my comment probably doesnt make sense